x/clerk/side_handler_test.go

```diff
@@ -0,0 +1,279 @@
+	t, ctx := suite.T(), suite.ctx
+
+	// post tx handler
+	_, err := suite.postHandler(ctx, nil, tmprototypes.SideTxResultType_YES)
+	require.Error(t, err)
+	// require.False(t, result.IsOK(), "Post handler should fail")
+	// require.Equal(t, sdk.CodeUnknownRequest, result.Code)
+}
+
+func (suite *SideHandlerTestSuite) TestPostHandleMsgEventRecord() {
+	t, app, ctx, r := suite.T(), suite.app, suite.ctx, suite.r
+
+	_, _, addr1 := testdata.KeyTestPubAddr()
+
+	id := r.Uint64()
+	logIndex := r.Uint64()
+	blockNumber := r.Uint64()
+	txHash := hmCommon.HexToHeimdallHash("no log hash")
+
+	msg := types.NewMsgEventRecord(
+		addr1,
+		txHash,
+		logIndex,
+		blockNumber,
+		id,
+		addr1,
+		make([]byte, 0),
+		suite.chainID,
+	)
+
+	t.Run("NoResult", func(t *testing.T) {
+		_, err := suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_NO)
+		require.Error(t, err)
+		// require.False(t, result.IsOK(), "Post handler should fail")
+		// require.Equal(t, common.CodeSideTxValidationFailed, result.Code)
+		// require.Equal(t, 0, len(result.Events), "No error should be emitted for failed post-tx")
+
+		// there should be no stored event record
+		storedEventRecord, err := app.ClerkKeeper.GetEventRecord(ctx, id)
+		require.Nil(t, storedEventRecord)
+		require.Error(t, err)
+	})
+
+	t.Run("YesResult", func(t *testing.T) {
+		_, err := suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_YES)
+		// require.True(t, result.IsOK(), "Post handler should succeed")
+		require.Nil(t, err)
+		// require.Greater(t, len(result.Events), 0, "Events should be emitted for successful post-tx")
+
+		// sequence id
+		blockNumber := new(big.Int).SetUint64(msg.BlockNumber)
+		sequence := new(big.Int).Mul(blockNumber, big.NewInt(hmCommon.DefaultLogIndexUnit))
+		sequence.Add(sequence, new(big.Int).SetUint64(msg.LogIndex))
+
+		// check sequence
+		hasSequence := app.ClerkKeeper.HasRecordSequence(ctx, sequence.String())
+		require.True(t, hasSequence, "Sequence should be stored correctly")
+
+		// there should be no stored event record
+		storedEventRecord, err := app.ClerkKeeper.GetEventRecord(ctx, id)
+		require.NotNil(t, storedEventRecord)
+		require.NoError(t, err)
+	})
+
+	t.Run("Replay", func(t *testing.T) {
+		id := r.Uint64()
+		logIndex := r.Uint64()
+		blockNumber := r.Uint64()
+		txHash := hmCommon.HexToHeimdallHash("Replay hash")
+		_, _, addr2 := testdata.KeyTestPubAddr()
+
+		msg := types.NewMsgEventRecord(
+			addr1,
+			txHash,
+			logIndex,
+			blockNumber,
+			id,
+			addr2,
+			make([]byte, 0),
+			suite.chainID,
+		)
+
+		_, err := suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_YES)
+		require.Nil(t, err)
+		// require.True(t, result.IsOK(), "Post handler should succeed")
+
+		_, err = suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_YES)
+		require.Error(t, err)
+		// require.False(t, result.IsOK(), "Post handler should prevent replay attack")
+		// require.Equal(t, common.CodeOldTx, result.Code)
```

Comment:
I feel we should check the following in all cases:
1. for error: (a) returned value should be nil, (b) error should be present, (c) error returned should be equated with the expected error
2. for success: (a) error should be nil, (b) returned value should be present, (c) if there's an predetermined return value, that should be equated with the actual value returned

Reply:
@ssandeep : any idea how to check for exact error?

Reply:
You can refer https://github.com/maticnetwork/heimdall/pull/519/files#diff-2e0ea1a4c73ff1bbc13d1079e303011edb4c62cd5e5533d859a667f1ceff8b9eR111